feat: Add surface code visualization demo

zzzappy opened the pull request:
## 📝 Description

Adds a demo notebook and associated code snippet to the docs which demonstrates the creation of surface code visualizations from `deltakit`-generated circuits using the `crumpy` and `plotly` packages.

This notebook was created as part of my Summer 2025 internship with Riverlane.

---

## 🚦 Status

Interactive widgets don't currently render in Sphinx - may want to move to different folder or figure out another way to display interactive visualizations.

---

## 🛠️ Future Work

`draw_patch` should be integrated with `CSSCode`.

---

## ➕️ Additional Information

My internship concludes tomorrow, so I will only be available for the next few days to respond.

---

## 🧾 Release Note

feat: Add surface code visualization demo

Commit:
Add plotly snippet

zzzappy committed:
Add initial surface_code_demo notebook

Still WIP: better text/narrative, better resizing for surface code patch visualizer

zzzappy committed:
surface_code_demo notebook updates

Refactor shared plotly/crumpy viz into widget, add +/- buttons to sliders, clean up explanatory text, other edits

zzzappy committed:
Update surface_code_demo.ipynb

Add imports section, refactor InteractivePlanarCode to update plotly graph in place with new update_patch in snippet.py (for smoother final output)

zzzappy committed:
Move surface_code_demo.ipynb and snippet to docs

zzzappy committed:
Remove images

mdhaber commented on docs/examples/notebooks/visualization/snippet.py:
```diff
@@ -0,0 +1,631 @@
+import matplotlib.pyplot as plt
+import numpy as np
+from deltakit_circuit import Qubit
+from deltakit_circuit.basic_types import Coord2D, Coord2DDelta
+from deltakit_circuit.qubit_identifiers import PauliGate, PauliX, PauliZ
+from deltakit_explorer.codes.css._css_code import CSSCode
+from deltakit_explorer.codes.stabiliser import Stabiliser
+from deltakit_explorer.enums.basic_enums import DrawingColours
```
I needed these to make this run with `main`. 
```suggestion
from deltakit_circuit._basic_types import Coord2D, Coord2DDelta
from deltakit_circuit._qubit_identifiers import PauliGate, PauliX, PauliZ
from deltakit_explorer.codes._css._css_code import CSSCode
from deltakit_explorer.codes._stabiliser import Stabiliser
from deltakit_explorer.enums._basic_enums import DrawingColours
```

Also, my IDE says there a lot of unused imports (gray).
<img width="597" height="563" alt="image" src="https://github.com/user-attachments/assets/df40c8b0-a5b7-4438-b029-00049ac65114" />

Can these be removed?

guenp replied:
That looks like a lot of private imports, would it make sense to use the public-facing imports instead?

mdhaber replied:
Yeah, good idea. I was going to mention it, too, but `DrawingColors` isn't public and a almost all thre rest can be removed. But after that, it would still be good to import `PauliX` from `deltakit.circuit` directly.

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+   "id": "856bf3cc",
+   "metadata": {},
+   "source": [
+    "We can add some interactivity with a slider (to change which qubit line the Pauli marker/error is on) and some radio buttons (to select the Pauli type):"
```
Super cool! But one thing I saw here and below:

<img width="179" height="274" alt="image" src="https://github.com/user-attachments/assets/dfa67563-a1c7-4307-b7fe-d5847c9118f4" />

Shouldn't `None` show the circuit, just without any errors?

zzzappy replied:
Good catch - missed a pair of parens in `circuitWithError`. The "none" option should now show the circuit without any errors for all the visualizations with the Pauli error slider.

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+   "source": [
+    "### Interactivity with `plotly`\n",
+    "\n",
+    "Let's explore another interactive visualization. `draw_patch` uses `plotly` to visualize a patch of a given surface code, similar to the static visualization at the beginning of this notebook. Clicking on qubits on the interactive plot will also cycle between Pauli errors on that qubits (none → Pauli X (red) → Pauli Y (green) → Pauli Z (blue)):"
```
Is there another color that could be used so that the big and small green circles are more visually distinct?

<img width="503" height="428" alt="image" src="https://github.com/user-attachments/assets/aaddd4bd-31a0-42d1-ab3d-0ff4db9665bf" />

zzzappy replied:
I'll check on the "ancilla" vs "auxiliary" terminology.

For the coloring, I agree the Pauli Y green is too close to the data qubit colors. I updated it to a lime green that stands out more while keeping the Pauli XYZ = RGB color scheme clear:
<img width="327" height="325" alt="image" src="https://github.com/user-attachments/assets/236ccf11-bda4-41f9-b2c2-303c3dfbc750" />

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+    }
+   ],
+   "source": [
+    "CircuitWidget.from_stim(stim_circuit)"
```
I was wondering what the colored dots are for?

<img width="570" height="305" alt="image" src="https://github.com/user-attachments/assets/4893bfcd-8bfe-4f44-b876-4ec2dff0c5c7" />

Also, not sure it's the right thing to do, but did you consider allowing `CircuitWidget` to accept a deltakit circuit directly?

nelimee replied:
Larger dots show the observable "sensitivity region" (in other words, portions on the circuit that may impact the observable value if an error happens there). Smaller ones are about detectors "sensitivity region".

zzzappy replied:
I added some more explanation on what the dots are before that cell, but I still remove them in subsequent cells just to reduce the visual clutter.

As for taking circuits directly from `deltakit` that should be totally doable in `crumpy` since `deltakit` already has `.to_stim_circuit()`. I may add that option to `crumpy` soon.

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+   "source": [
+    "# Interactive Surface Codes with `deltakit` and `crumpy`\n",
+    "\n",
+    "In this notebook, we show how `deltakit` can be used to generate surface code memory experiments and various ways in which visualization packages like `crumpy` and `plotly` can be used to create interactive interfaces for surface code exploration. We show examples of generating planar codes with `deltakit.explorer.codes`, how to interactively visualize surface code experiment circuits with ` ...
```
Maybe link to `crumpy` and `plotly` websites?

zzzappy replied:
I added [links to their respective PyPI packages](https://github.com/Deltakit/deltakit/pull/28/files#diff-1b37fe26e7efcd5753017e5921c000831d07407e55aa3a2e657f976be23f618dR10)

mdhaber commented on docs/examples/index.rst:
```diff
@@ -20,3 +20,4 @@ let's take a look at some realistic examples.
    notebooks/simulation/server_simulation
    notebooks/simulation/leakage-aware_decoding
    notebooks/simulation/leakage_demo
+   notebooks/visualization/surface_code_demo
```
After fixing imports, please add this notebook to the list here so it runs in CI:
https://github.com/Deltakit/deltakit/blob/717fc7f4e7110d95e25f5793838b0d94f5f9b2a9/.github/workflows/run_notebooks.yml#L10-L20

zzzappy replied:
Added.

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+    "from deltakit.circuit import GateLayer, Circuit\n",
+    "from deltakit.circuit.gates import PauliBasis\n",
+    "from crumpy import CircuitWidget\n",
+    "from snippet import draw_patch, update_patch\n",
```
Is the plan to eliminate `snippet` and incorporate these things into `deltakit`?

zzzappy replied:
I'd have to check with Guen on this but I think that adding some form of interactive `plotly` visualization like `draw_patch` in `snippet.py` may be planned. She had worked up an initial interactive version of `draw_patch` (built off of the existing static `draw_patch` of `PlanarCode`) but changes like the RGB cycling on qubit clicks were things I added for this demo in particular. `update_patch` is a bit more rough but could also be useful in some form down the line.

guenp replied:
I would suggest creating a new module in `deltakit_explorer.visualisation` called `_planar_code.py` and move both `draw_patch` methods there. We can do that as part of this PR, or open a follow up PR after this one. @mdhaber , curious to hear your thoughts on this.

mdhaber reviewed (commented):
This is super cool! Amazing that you've gotten this to work so well so fast. Here are a few thoughts I had while going through this locally.

zzzappy committed:
Code review changes

- Still need to check on ancilla/auxiliary naming and fix mypy pre-commit
issue
- Also added a ruff.toml config to
  docs/examples/notebooks/visualizations (to ignore some rules in
  snippet.py to pass pre-commit)

zzzappy committed:
Fix DEP001 snippet ignore

zzzappy commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,1289 @@
+   "id": "5b9f24c4",
+   "metadata": {},
+   "source": [
+    "Using our `stimSurfaceCode` circuit generation function, let's explore another feature of `crumpy` and add a Pauli marker to a circuit:"
```
I added some more explanation under the Pauli Propagation section on what is happening in the Pauli marker circuit. For those less familiar with `crumpy`/Crumble hopefully the updated explanation gives a clearer understanding of what a Pauli marker is and why it's useful.

Having an easy way to insert errors at specific spots without having to go down to the level of string manipulation is something `crumpy` may want to provide in the future, but currently you do have to convert down to a string and write Stim instructions with special `#!pragma MARK` instructions for Pauli markers. There are `TICK` instructions in Stim that let you explicitly add timesteps, so you could so something like this:

```
circuit = """
R 0 1
TICK
CX 0 1
TICK
H 1
#!pragma MARKX(0) 1
TICK
M 1
"""
CircuitWidget(stim=circuit)
```
<img width="160" height="65" alt="image" src="https://github.com/user-attachments/assets/49b1a046-c91f-4a16-a903-4b8279fcc9b5" />

guenp commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,587 @@
+   "id": "5b9f24c4",
+   "metadata": {},
+   "source": [
+    "Using our `stimSurfaceCode` circuit generation function, let's explore another feature of `crumpy` and add a Pauli marker to a circuit.\n",
```
```suggestion
    "Using the `stimSurfaceCode` circuit generation function, let's explore another feature of `crumpy` and add a Pauli marker to a circuit.\n",
```

guenp commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,587 @@
+   "id": "15ab499e-9748-4221-8086-bed93b534cc1",
+   "metadata": {},
+   "source": [
+    "By linking this widget with a `CircuitWidget`, we can create an clickable surface code patch with a linked circuit visualization. Clicking a qubit on the interactive graph changes the Pauli error present at that qubit on the circuit diagram (with errors automatically propagating through the circuit):"
```
```suggestion
    "By linking this widget with a `CircuitWidget`, you can create an clickable surface code patch with a linked circuit visualization. Clicking a qubit on the interactive graph changes the Pauli error present at that qubit on the circuit diagram (with errors automatically propagating through the circuit):"
```

guenp reviewed (approved):
Nice work, @zzzappy ! I've made some suggestions to make the language use the active voice and "you" vs. "we", inspired by https://developers.google.com/style/voice and https://developers.google.com/style/person.

mdhaber reviewed (commented):
Given https://github.com/Deltakit/deltakit/pull/28#pullrequestreview-3260260173, these are the the other changes I see to keep the voice consistent (found by searching for "we" and "our").

mdhaber commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,587 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from deltakit.explorer.codes import RotatedPlanarCode, UnrotatedPlanarCode, css_code_memory_circuit\n",
```
This is not specific to this notebook, but I realize that the default line length before a scroll bar gets added is 79 characters.

<img width="738" height="318" alt="image" src="https://github.com/user-attachments/assets/357c973f-796c-4d9b-80d3-026ba5665cde" />

No need to change anything now; just an observation.

guenp commented on docs/examples/notebooks/visualization/surface_code_demo.ipynb:
```diff
@@ -0,0 +1,587 @@
+   "id": "8112ffee",
+   "metadata": {},
+   "source": [
+    "# Interactive Surface Codes with `deltakit` and `crumpy`\n",
```
Yes, that's right, capitalizing it in a header or sentence is preferred, unless you're talking about the package in particular (e.g. "Run `import deltakit`" or "Install `deltakit`").

guenp committed:
Apply suggestions from code review

Co-authored-by: Matt Haberland <matt.haberland@gmail.com>

guenp committed:
Merge branch 'main' into zzzappy/crumpy-surface-code-demo

nelimee commented:
Comment on the state of this PR: I will review it / make the necessary changes to include it, but don't have the time to do it for the moment.

rolandriver commented:
Closing as no follow-up work on this. Re-open if necessary.
